### ISSUE TITLE
Serve LLM models

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/Blogs#56]
A few options to explore
1. NVIDIA NeMo, TensorRT_LLM, Triton

- NeMo

Run [this Generative AI example](https://github.com/NVIDIA/GenerativeAIExamples/tree/main/models/Gemma
) to build Lora with Gemma 2b and 7b 

Run [this example](https://docs.nvidia.com/nemo-framework/user-guide/latest/playbooks/mistralpeft.html) with Mistral 7B 
for PEFT

- TensorRT/TensorRT_LLM

 
- Triton Inference Server

2. vLLM
This [example ](https://github.com/meta-llama/llama-recipes/tree/main/recipes/inference/model_servers
)shows how to use LLAMA and vLLM to build an app.

**Comment by YunchaoYang** [comment]
## OpenLLM
[星策社区一站式LLMops meetup](https://www.bilibili.com/video/BV1GN411h7Vk/?spm_id_from=333.337.search-card.all.click&vd_source=47dc985a28b3fde2f1c87fb9f0d91d29)

## LiteLLM
similar to langchain, unified API, 
[litellm 统一大模型接口](https://www.bilibili.com/video/BV1Wx4y1S7qs/?spm_id_from=333.337.search-card.all.click&vd_source=47dc985a28b3fde2f1c87fb9f0d91d29)

**Comment by YunchaoYang** [comment]
## Trtion Inference Server

- [原理入门之框架篇](https://www.bilibili.com/video/BV1KS4y1v7zd/?spm_id_from=333.337.search-card.all.click&vd_source=47dc985a28b3fde2f1c87fb9f0d91d29)

- [原理入门之产品篇](https://www.bilibili.com/video/BV11Z4y1q7vW/?spm_id_from=333.788&vd_source=47dc985a28b3fde2f1c87fb9f0d91d29)

### Handson
- [上手实战教程](https://www.bilibili.com/video/BV1234y157Xh/?spm_id_from=333.788&vd_source=47dc985a28b3fde2f1c87fb9f0d91d29)

#### 1. Prepare the model repo
<img width="454" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/ffd0faa5-0108-4165-b1a0-2cc0621ac57b">
<img width="424" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/f7753af7-fadb-4b9c-826e-d7fedacd4382">

#### 2. config model
the minimal necessary parameters configure served model
1. platform/backend: to define which backend to use 2. max_batch_size 3. input and output

For tensorRT, onnx file, tensorflow saved-model, do not require config.pbtxt when --strict-model-config=False.
platform/backend: 用于指定后端，大部分情况二选一，特殊情况需要特殊对待，见后面。

max_batch_size: 指定最大batch。

input、output: 输入输出Tensor的名字和信息。

注意，对于Tensorrt，TensorFlow save-model，onnx模型，config.pbtxt不是必须的，只要启动指定 --strict-model-config=false

<img width="589" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/d453e8a0-b765-4877-8999-aa0fbc8a3081">

<img width="678" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/74b3f5d4-2f4b-4e56-8e30-fb3a1d8cc850">
对于Tensorrt、onnxrt、pytorch，这两种参数二选一即可。

对于TensorFlow必须指定platform，backend可选。

对于openvino，python，dali，只能使用backend。

对于Custom，21.05版本之前，可以通过platform参数设置为custom表示；之后必须通过backend字段进行指定，值为你的custom backend library的名字。

The version policy in config.pbtxt, you can decide which one to serve.

<img width="529" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/f73b0941-531b-427d-8244-28f01ae065e0">
情况1：

max_batch_size为一个大于0的常数，Input和output指定名字，数据类型，数据形状。

注意：dims在指定的时候忽略batch_size的维度。

情况2：

max_batch_size等于0。表示模型的输入和输出是不包括batch_size那个维度的。

这个时候维度信息就是真实的维度信息。

情况3：

pytorch特殊情况，torchscript模型不保存输入输出的名字，因此对输入输出名称有特殊规定，"字符串__数字"。

支持可变shape，设置为-1。

情况4：

reshape参数：对输入输出进行reshape。


version_policy参数，策略：

all：加载所有版本的模型。

latest：加载最新的模型（可多个，版本号越大越新）

specific：指定特定的版本。

五、Instance Groups




对应triton的并行计算能力特性，这个参数主要用来配置在指定设备上运行多个实例，提高模型服务能力，增加吞吐。

Instance Groups配置跑在同样设备上的一组模型实例。

count：同时开启的模型数量。

kind：指定设备类型。

gpus：指定GPU编号，如果不指定这个参数，triton会在每个GPU上跑相应数量的instance。

可配置多组。

对应triton的并行计算能力特性，这个参数主要用来配置在指定设备上运行多个实例，提高模型服务能力，增加吞吐。

Instance Groups配置跑在同样设备上的一组模型实例。

count：同时开启的模型数量。

kind：指定设备类型。

gpus：指定GPU编号，如果不指定这个参数，triton会在每个GPU上跑相应数量的instance。

可配置多组。

六、Scheduling And Batching

Scheduling：指定调度策略来应对请求。

6.1 Default Scheduler

不做batching；
输入进来是多少就按照多少去推理；


6.2 Dynamic Batcher

在服务端将多少个batch_size比较小的input_tensor合并为一个batch_size比较大的input_tensor；
提高吞吐率的关键手段；
只适合无状态模型；
子参数：

preferr_batch_size: 期望达到的batch_size是多少，多个值；

max_queue_delay_microseconds: 打成batch的时间限制，微秒；

高级子参数：

preserver_ordering: 请求进来的顺序和响应出去的顺序保持一致；

priority_levels: 定义不同优先级请求处理顺序；

Queue_Policy: 设置请求等待队列行为；

6.3 Sequence Batcher

专门用于stateful model的一种调度器；
确保同一序列的推理请求能够路由到同样的模型实例上推理；
6.4 Ensemble Scheduler

组合不同的模块，形成pipeline；
后面详细介绍。
七、Optimization Policy

Onnx模型优化——TRT backend for ONNX；
TensorFlow模型优化——TF-TRT；
八、Model Warmup

指定模型热身的参数；

初始化可能延迟，直到收到前面几个推理请求；
热身完成后，Triton的服务才是Ready状态；
模型加载会变长；

**Comment by YunchaoYang** [comment]
![image](https://github.com/YunchaoYang/Blogs/assets/6526592/c04dfa4c-97cb-450d-b4f0-8af339782309)
1. 开启Triton容器
2. 运行Triton服务
```
tritonserver --model-repository=<MODEL_REPO> 
                   --log-verbose
                    --strict-model-config
                   --strict-readiness
                   --exit-on-error <boolean>
                   --http-port <integer>
                   --grpc-port <integer>
                   --metrics-port <integer>
                   --model-control-mode <string>
```
### 检查服务是否准备就绪
`curl -v <IP>:8000/v2/health/ready`

### 其他启动选项
--log-verbose <integer>: 开启verbose日志信息。0 or >=1 
--strict-model-config <boolean>: 是否需要配置模型。true of false
--strict-readiness <boolean>: ready状态显示状况。
--exit-on-error <boolean>: 模型加载部分失败，是否也启动。
--http-port <integer>: 指定http服务端口，默认是8000。
--grpc-port <integer>: 指定GRPC服务端口，默认是8001。
--metrics-port <integer>: metrics报告端口，默认8002。
--model-control-mode <string>: 模型管理模式，默认是none，把模型库中所有模型都load进来，并且无法动态卸载或者更新。explicit，server启动时不加载模型，可以通过api进行加载或者卸载模型；poll，动态更新模型，增加新的版本或者修改配置，服务都会动态去加载模型。

#### other options
--repository-poll-secs <integer>: 模型控制模式为poll时，自动检查模型库变动的时间。
--load-model <string>: 模型控制模式为explicit时指定启动时加载的模型。
--pinned-memory-pool-byte-size <integer>：可以被Triton服务使用的锁页内存大小，关于锁页内存可以参考：https://cloud.tencent.com/developer/article/2000487。Default 256MB
--cuda-memory-pool-byte-size <<integer>:<integer>>：可以被Triton使用的cuda memory 大小。default 64MB
--backend-directory <string>: backend搜索路径，可在使用custom backend的时候指定自己的库。
--repoagent-directory <string>：预处理模型库的库，譬如在load模型的时候进行加密。



```
curl -X POST http://localhost:8000/v2/repository/models/resnet50_pytorch/load
curl -X POST http://localhost:8000/v2/repository/models/resnet50_pytorch/unload

```

**Comment by YunchaoYang** [comment]
# Configure an Ensemble Model (Pipeline)
<img width="476" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/0b91fded-93b4-4fb9-91cb-cfb6d8d56e33">

<img width="671" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/9adcbff5-cc11-4afe-b78f-dfa657d2afd3">
#### ensemble notes
1. name matches of the output and input of sequential models
2. if model is stateful, need to container information in inference request
3. The model composing the ensemble have their own scheduler
4. If Model in ensember are all framework backends, data transmission between them does not have to go through GPU memory.

#### Example 2 ASR

**Comment by YunchaoYang** [comment]
# client 发送请求到Triton服务

## 同步方式推理：
![image](https://github.com/YunchaoYang/Blogs/assets/6526592/ffebdb47-2e14-4d9f-ba73-a7317c2fe5b6)
## 异步方式推理：
![image](https://github.com/YunchaoYang/Blogs/assets/6526592/0b23000b-52af-46ac-94fa-37e6666443a0)
通过share memory去推理：适合client和server在同一部机器上的情况。
![image](https://github.com/YunchaoYang/Blogs/assets/6526592/1387c75c-67ad-4681-9eec-c133d813000f)

**Comment by YunchaoYang** [comment]
# TensorRT-LLM

https://www.bilibili.com/video/BV1Xu4m1g74A/?spm_id_from=autoNext&vd_source=47dc985a28b3fde2f1c87fb9f0d91d29

<img width="437" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/18b939a3-d410-4dcc-8033-812b2625af27">
<img width="332" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/4ef7e34a-f089-4275-beeb-2cc1574e39f2">
<img width="400" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/9240f5f4-4e97-4b37-8b08-985c13501bad">
<img width="400" alt="image" src="https://github.com/YunchaoYang/Blogs/assets/6526592/abc48aeb-3b47-4222-93d2-e813a925d468">

**Comment by YunchaoYang** [comment]
More reference videos:

https://www.youtube.com/watch?v=P7dvC31Ggxk
https://www.youtube.com/watch?v=BuJzbgExUTQ
https://www.youtube.com/watch?v=U2WRg2Rlg1o
https://www.youtube.com/watch?v=NR_iUl2Ooc0
